### PR TITLE
Optionally use PnetCDF independent data mode in SCORPIO

### PR DESCRIPTION
**Pull request by dqwu**
Entering and leaving PnetCDF independent data mode is expensive.
Wei-keng Liao (principal PnetCDF developer) has provided a patch
to avoid that by making non-root processes do zero-sized reads
while still in collective mode.

Based on Wei-keng's patch, this PR added a compile time configure
option to control if SCORPIO uses PnetCDF independent data mode.

On E3SM machine Chrysalis, this option is turned OFF by default as
entering independent data mode might fail with Open MPI 4.1.2 or
older versions during parallel read of a NetCDF file.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -179,6 +180,20 @@ else()
...
+    message(STATUS "Using PnetCDF collective data mode to read variables in SCORPIO on E3SM machine Chrysalis")
+  else()
+    set(USE_INDEP_MODE 1)
+    message(STATUS "Using PnetCDF independent data mode to read variables in SCORPIO (default)")
+  endif()
+else()
+  set(USE_INDEP_MODE 0)
+  message(STATUS "Using PnetCDF collective data mode to read variables in SCORPIO")
+endif()
+
```
[the remark this answers is not in the document]
Comment (dqwu) [reply CMakeLists.txt:196]: OK, this has been updated. Besides Chrysalis, the option is not explicitly enabled for ANL GCE nodes (to test new code on collective data mode).

File: src/clib/pio_getput_int.c [hunk 2]
```diff
@@ -1025,6 +1026,52 @@ int PIOc_get_vars_tc(int ncid, int varid, const PIO_Offset *start, const PIO_Off
                 return pio_err(ios, file, ierr, __FILE__, __LINE__,
                                 "Reading variable (%s, varid=%d) from file (%s, ncid=%d) failed. Ending independent (across processes) access failed on the file", pio_get_vname_from_file(file, varid), varid, pio_get_vname_from_file(file, varid), ncid);
             }
+#else
+            int var_ndims;
+            PIO_Offset *cnt;
+
+            ncmpi_inq_varndims(file->fh, varid, &var_ndims);
```
[the remark this answers is not in the document]
Comment (dqwu) [reply src/clib/pio_getput_int.c]: Actually we do not need to call ncmpi_inq_varndims to get var_ndims here. An existing variable ndims can be used instead.

File: CMakeLists.txt [hunk 3]
```diff
@@ -179,6 +180,20 @@ else()
   endif()
 endif()
 
+if(PIO_USE_INDEP_MODE)
+  cmake_host_system_information (RESULT FQDN_SITENAME QUERY FQDN)
+  if(FQDN_SITENAME MATCHES "^chrlogin")
+    set(USE_INDEP_MODE 0)
+    message(STATUS "Using PnetCDF collective data mode to read variables in SCORPIO on E3SM machine Chrysalis")
```
[the remark this answers is not in the document]
Comment (dqwu) [reply CMakeLists.txt]: OK, it has been removed.

File: src/clib/pio_getput_int.c [hunk 4]
```diff
@@ -1025,6 +1026,52 @@ int PIOc_get_vars_tc(int ncid, int varid, const PIO_Offset *start, const PIO_Off
...
+            if (ios->iomaster == MPI_ROOT)
+                cnt = (PIO_Offset*)count;
+            else
+                /* non-root reads nothing */
+                cnt = (PIO_Offset*) calloc(var_ndims, sizeof(PIO_Offset));
```
[the remark this answers is not in the document]
Comment (dqwu) [reply src/clib/pio_getput_int.c]: OK, this has been updated.

**Review by jayeshkrishna — APPROVED** [review]

**Pull request merged**